.github/workflows/ci-cd.yml

```diff
@@ -0,0 +1,120 @@
+---
+name: CI/CD
+
+on:
+  push:
+    branches: [master]
+  pull_request:
+    branches: [master]
+  workflow_dispatch:
+
+env:
+  FORCE_COLOR: "1" # Make tools pretty.
+  PIP_DISABLE_PIP_VERSION_CHECK: "1"
+  PIP_NO_PYTHON_VERSION_WARNING: "1"
+  PYTHON_LATEST: "3.11"
+
+  # For re-actors/checkout-python-sdist
+  sdist-artifact: python-package-distributions
+
+permissions:
+  contents: read
+
+jobs:
+  build-sdist:
+    name: 📦 Build the source distribution
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v3
+        name: Checkout project
+        with:
+          fetch-depth: 0
+      - uses: actions/setup-python@v4
+        name: Setup ${{ env.PYTHON_LATEST }}
+        with:
+          python-version: ${{ env.PYTHON_LATEST }}
+      - run: python -Im pip install build
+        name: Install core libraries for build and install
+
+      - run: python -Im build
+        name: Build artifacts
+
+      - uses: actions/upload-artifact@v3
+        name: Upload built artifacts for testing
+        with:
+          name: ${{ env.sdist-artifact }}
+          # NOTE: Exact expected file names are specified here
+          # NOTE: as a safety measure — if anything weird ends
+          # NOTE: up being in this dir or not all dists will be
+          # NOTE: produced, this will fail the workflow.
+          path: dist/${{ env.sdist-name }}
+          retention-days: 15
+
+  tests:
+    name: Tests on ${{ matrix.python-version }}
+    needs: build-sdist
+    runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        python-version:
+          - "3.8"
+          - "3.9"
+          - "3.10"
+          - "3.11"
+          - "~3.12.0-0"
+    continue-on-error: >-
+      ${{ contains(matrix.python-version, '~') && true || false }}
+    steps:
+      - name: Get source code from pre-built sdist
+        uses: re-actors/checkout-python-sdist@release/v1
+        with:
+          workflow-artifact-name: ${{ env.sdist-artifact }}
+      - uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python-version }}
+      - uses: actions/checkout@v3
+        name: Checkout project to get necessary dependencies for testing
+        with:
+          fetch-depth: 0
+      - name: Install dependencies
+        run: |
+          pip install --upgrade setuptools
+          pip install -r requirements-dev.txt
```

Comment:
If you know a more efficient way to install these requirements without needing to checkout the project, please let me know.

Reply:
I think it's not only fine but also required to checkout the project to run tests.